examples/machine_translation/char_dataset.py

```diff
@@ -0,0 +1,121 @@
+import itertools
+import re
+
+import torch
+from torch.utils.data import DataLoader
+
+from torchtext.data.utils import get_tokenizer
+from torchtext.experimental.datasets.translation import (DATASETS,
+                                                         TranslationDataset)
+from torchtext.experimental.functional import sequential_transforms, vocab_func
+from torchtext.vocab import build_vocab_from_iterator
+
+
+def build_word_vocab(data, transforms, index, init_token="<w>", eos_token="</w>"):
+    tok_list = [[init_token], [eos_token]]
+    for line in data:
+        tok_list.append(transforms(line[index]))
+    return build_vocab_from_iterator(tok_list)
+
+
+def build_char_vocab(
+    data, transforms, index, init_word_token="<w>", eos_word_token="</w>", init_sent_token="<s>", eos_sent_token="</s>",
+):
+    tok_list = [
+        [init_word_token],
+        [eos_word_token],
+        [init_sent_token],
+        [eos_sent_token],
+    ]
+    for line in data:
+        tokens = list(itertools.chain.from_iterable(transforms(line[index])))
+        tok_list.append(tokens)
+    return build_vocab_from_iterator(tok_list)
+
+
+def char_vocab_func(vocab):
```

Comment:
Is this something like subword method? Do you think if sentencepiece method is easier and more helpful?

Reply:
By subword method you mean tokenizing sentence into a list of subwords? Then no. This is a function with similar functionality with `vocab_func`, converting a list of strings to their index, except instead of converting words to string we are converting a list of chars to their index. To give you an example


```
sentence: I love torchtext
character representation: [['I'], ['l', 'o', 'v', 'e'], ['t', 'o', 'r', 'c', 'h', 't', 'e', 'x', 't']]
char_vocab_func(character representation): [[1], [2, 3, 4], [5, 6, 7, 8, 9, 10, 11, 12, 13]]
```